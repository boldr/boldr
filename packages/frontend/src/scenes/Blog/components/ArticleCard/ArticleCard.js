/* @flow */
import React from 'react';
import { format } from 'date-fns';
import Link from 'react-router-dom/Link';
import classnames from 'classnames';
import { Button, Card, CardTitle, CardText, CardActions, Media, StyleClasses } from '@boldr/ui';
import type { Article } from '../../../../types/boldr';

import TagBlock from '../TagBlock';

const BASE_ELEMENT = StyleClasses.ARTICLE_CARD;

type Props = {
  className?: string,
  article: Article,
};

const ArticleCard = (props: Props) => {
  const formattedDate = format(props.article.createdAt, 'MM/DD/YYYY');
  const classes = classnames(BASE_ELEMENT, props.className);
  const { title, image, slug, tags, excerpt } = props.article;
  return (
    <div className={classes}>
      <Card style={{ maxWidth: 600 }} className="md-block-centered">
        <Media>
          <img src={image} alt={`${title} feature image`} role="presentation" />
        </Media>
        <CardTitle title={title} subtitle={formattedDate} />
        <CardActions expander>
          <Link to={`/blog/${slug}`}>
            <Button kind="primary" outline>
              Read More
            </Button>
          </Link>
        </CardActions>
        <CardText expandable>{excerpt}</CardText>
        <TagBlock tags={tags} />
      </Card>
    </div>
  );
};

export default ArticleCard;
