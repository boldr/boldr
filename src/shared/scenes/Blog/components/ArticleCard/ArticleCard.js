/* @flow */
import React from 'react';
import Link from 'react-router-dom/Link';
import { connect } from 'react-redux';
import dateFns from 'date-fns';
import classnames from 'classnames';
import FlatButton from 'material-ui/FlatButton';
import { Col, Row, StyleClasses } from 'boldr-ui';
import {
  Card,
  CardActions,
  CardMedia,
  CardTitle,
  CardText,
} from 'material-ui/Card';
import { selectArticle } from '../../state/articles/actions';

import TagBlock from '../TagBlock';

const BASE_ELEMENT = StyleClasses.POST_CARD;
type Props = {
  className: string,
  id: ?string,
  featureImage: ?string,
  title: string,
  slug: string,
  content: ?string,
  backgroundImage: ?string,
  excerpt: ?string,
  createdAt: string,
  updatedAt: ?string,
  status: ?string,
  author: string,
  seo: ?Object,
  tags: ?Array<Tag>,
  attachments: ?Object,
  meta: ?Object,
  userId: ?string,
  dispatch: Function,
  listTags: Object,
};

export const ArticleCard = (props: Props) => {
  const formattedDate = dateFns.format(props.createdAt, 'MM/DD/YYYY');
  const classes = classnames(BASE_ELEMENT, props.className);
  const articleTags = props.tags
    ? props.tags.map(id => props.listTags[id])
    : null;
  // Explicitly define post rather than passing additional
  // unnecessary props like listTags
  const article = {
    id: props.id,
    author: props.author,
    attachments: props.attachments,
    content: props.content,
    createdAt: props.createdAt,
    excerpt: props.excerpt,
    backgroundImage: props.backgroundImage,
    featureImage: props.featureImage,
    meta: props.meta,
    slug: props.slug,
    status: props.status,
    tags: props.tags,
    title: props.title,
    userId: props.userId,
  };
  function transitionPost() {
    props.dispatch(selectArticle(article));
  }

  return (
    <div className={classes}>
      <Card>
        <CardMedia
          overlay={<CardTitle title={props.title} subtitle={formattedDate} />}
        >
          <img src={props.featureImage} alt={`${props.title} feature image`} />
        </CardMedia>

        <CardText>
          {props.excerpt}
          <Row>
            <Col xs={12}>
              <Link to={`/blog/${props.slug}`} className="readmore-link">
                <FlatButton
                  primary
                  label="Read More"
                  onTouchTap={transitionPost}
                />
              </Link>
            </Col>
          </Row>
        </CardText>
        <CardActions>
          <TagBlock tags={articleTags} />
        </CardActions>
      </Card>
    </div>
  );
};

export default connect(state => state, null, null, { pure: true })(ArticleCard);
