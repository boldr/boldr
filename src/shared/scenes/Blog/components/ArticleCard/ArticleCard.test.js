import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { ArticleCard } from './ArticleCard';

describe('<ArticleCard />', () => {
  it('renders <ArticleCard /> without breaking', () => {
    const wrapper = shallow(<ArticleCard title="blah" />);
    expect(wrapper.find('div').length).toBe(1);
  });

  it('accepts props and renders them.', () => {
    const wrapper = shallow(
      <ArticleCard
        feature_image="http://boldr.io/images/logo.png"
        title="Test Post"
        excerpt="An excerpt"
        slug="test-post"
      />,
    );
    expect(wrapper.instance().props.feature_image).toBe(
      'http://boldr.io/images/logo.png',
    );
    expect(wrapper.instance().props.title).toBe('Test Post');
    expect(wrapper.instance().props.excerpt).toBe('An excerpt');
    expect(wrapper.instance().props.slug).toBe('test-post');
  });
  it('renders snapshot', () => {
    const wrapper = shallow(<ArticleCard title="Test Post" />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
