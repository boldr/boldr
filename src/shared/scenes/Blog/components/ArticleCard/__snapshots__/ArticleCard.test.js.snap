// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleCard /> renders snapshot 1`] = `
<div
  className="boldrui-post-card"
>
  <Card
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
  >
    <CardMedia
      overlay={
        <CardTitle
          subtitle="Invalid Date"
          title="Test Post"
        />
      }
    >
      <img
        alt="Test Post feature image"
      />
    </CardMedia>
    <CardText>
      <Row
        componentClass="div"
      >
        <Col
          componentClass="div"
          xs={12}
        >
          <Link
            className="readmore-link"
            replace={false}
            to="/blog/undefined"
          >
            <FlatButton
              disabled={false}
              fullWidth={false}
              label="Read More"
              labelPosition="after"
              labelStyle={Object {}}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              onTouchTap={[Function]}
              primary={true}
              secondary={false}
            />
          </Link>
        </Col>
      </Row>
    </CardText>
    <CardActions>
      <TagBlock
        tags={null}
      />
    </CardActions>
  </Card>
</div>
`;
